### PR TITLE
feat(regex): use libpcre2

### PR DESCRIPTION
**Pull request by kopecs**
This is another go around for https://github.com/semgrep/semgrep/pull/9919 (description copied below), with some additional changes around how libpcre2 is linked. Uses linking updates from https://github.com/semgrep/semgrep/pull/10122.
---

We currently use libpcre (aka PCRE; libpcre3 in Debian) for our user-facing regex engine (Internally, `Re` and `Str` are also used in some cases). libpcre is end of life, and has been since at least 2022. Additionally, the successor libpcre2 has been actively developed since ~2015 and has a fair amount of improvements over libpcre 8.45. Specifically, this would use libpcre2-8 (i.e., the 8 bit/UTF-8 runtime).

This would migrate the regex language, but _not_ the internal matching for generic, to use libpcre2. Generic is not included since:
- There are a significant number of test failures on libpcre2-8 10.34 (the version which Debian distributes by default); libpcre2-8 10.43 is the latest version and appears to resolve most issues here.
- Updating generic mode is a reasonable amount of work on top of what this PR does already.

Additionally, this combines `Regex_engine` into the PCRE modules.

For reviewers,  the most important files are:
 - `src/commons/{Legacy_,}Regex.ml{,i}`
 - `js/libpcre2/runtime.js`

Most other changes are either build system modifications (e.g., add `dev-libpcre2`) or renaming modules in other files.

See also https://www.notion.so/semgrep/Migrate-from-PCRE-to-PCRE2-24e6d9f5b1db47818fe44d51508bc11f?pvs=4.

Note: I have added pcre2-ocaml as a submodule for now based on conversation with Martin, since the current bindings have at least one issue (fixed in the vendored version) relating to the creation of ocaml strings used for error messages with `BadPattern` (See [`semgrep/pcre2-ocaml@aea672`](https://github.com/semgrep/pcre2-ocaml/commit/aea672291df0139b341008f359e7188aca6081ca)).

Test plan: `make test && make test-jsoo`. Includes existing coverage for matching logic; some duplicate tests created for `Pcre2_`, copied from those we had for `Pcre_`; net-new tests for the js bindings based on the libpcre bindings.

BREAKING CHANGE: Migration from libpcre to libpcre2 changes slightly the regex we accept. As an example, `[\w-_]` would now be rejected (either `-` would need to be escaped, or it placed at the end, to be clear it is not a character range).

### CODE REVIEW

**Comment by kopecs** [comment]
Hmm, looks like it isn't present on x86, but is on arm?

File: Makefile [hunk 1]
```diff
@@ -349,27 +349,52 @@ install-deps: install-deps-for-semgrep-core
...
+					 gmp-dev \
+					 libev-dev \
+					 curl-dev \
+					 openssl-libs-static \
+					 zlib-static
 
 # Here is why we need those external packages below for pysemgrep:
 # - python3: obviously needed for pysemgrep and our e2e tests
-ALPINE_APK_DEPS_PYSEMGREP=python3
+# - python-dev: for compiling jsonnet for pysemgrep
```
Comment (aryx) [inline Makefile]: you can remove python3-dev here. it's not in develop anymore because we don't support jsonnet for pysemgrep anymore.

File: Makefile [hunk 2]
```diff
@@ -349,27 +349,52 @@ install-deps: install-deps-for-semgrep-core
...
+ALPINE_APK_DEPS_PYSEMGREP=python3 \
+						  python3-dev
 
 # Here is why we need those external packages:
 # - pkg-config?
-UBUNTU_DEPS=pkg-config libpcre3-dev libgmp-dev libev-dev libcurl4-gnutls-dev
+# NOTE: libpcre3 is actually libpcre
```
Comment (aryx) [inline Makefile:380]: lol

File: changelog.d/scrt-467.changed [hunk 3]
```diff
@@ -0,0 +1,5 @@
+The main regex engine is now PCRE2 (was PCRE). While the syntax is mostly
+compatible, there are some minor instances where updates to rules may be
+needed, since PCRE2 is slightly more strict in some cases. For example, while
+we previously accepted `[\w-.]`, such a pattern would now need to be written
```
Comment (aryx) [inline changelog.d/scrt-467.changed:4]: what [\w-.] meant? any word charater or '-' or '.'?
and what [\w_.] mean? any word character or '_' or '.'?
why this '_' suddenly?

File: dune-project [hunk 4]
```diff
@@ -104,6 +104,7 @@ a few other projects developed at r2c.
     (yojson (>= "1.7.0"))
     (re (>= "1.10.4"))
     (pcre (>= "7.5.0" ))
+    (pcre2 (>= "7.5.2" ))
```
Comment (aryx) [inline dune-project]: is it the version we use in our pcre2-ocaml fork? Then maybe we should bump the version?

File: js/libpcre/runtime.js [hunk 5]
```diff
@@ -167,7 +167,7 @@ function pcre_alloc_string(js_string) {
   var ptr;
   const array = libpcre.intArrayFromString(js_string);
   const length = array.length;
-  ptr = libpcre._malloc(length + 1);
+  ptr = libpcre._malloc(length);
```
Comment (aryx) [inline js/libpcre/runtime.js]: seems unrelated to this PR, maybe we should have those fixes in a separate PR?
or did you merge the latest develop?

File: libs/aliengrep/Match.ml [hunk 6]
```diff
@@ -1,6 +1,7 @@
 (*
    Match a compiled pattern against a target string.
 *)
+[@@@alert "-deprecated"]
```
Comment (aryx) [inline libs/aliengrep/Match.ml:8]: Why this directive?

File: libs/commons/Pcre2_.ml [hunk 7]
```diff
@@ -0,0 +1,344 @@
+(* Yoann Padioleau, Cooper Pierce, Martin Jambon
```
Comment (aryx) [inline libs/commons/Pcre2_.ml:1]: If it's mostly a copy paste of Pcre_.ml, then say it early, and maybe explain
what are the differences.

File: libs/spacegrep/src/lib/Comment.ml [hunk 8]
```diff
@@ -1,6 +1,7 @@
 (*
    Strip comments and the like as a preprocessing phase.
 *)
+[@@@alert "-deprecated"]
```
Comment (aryx) [inline libs/spacegrep/src/lib/Comment.ml:8]: maybe add a comment explaining it's because of the use of Pcre_ that should be migrated to Pcre2_

File: scripts/osx-setup-for-release.sh [hunk 9]
```diff
@@ -6,14 +6,36 @@ set -eux
 # history: there used to be a separate osx-m1-release.sh script
 # that was mostly a copy of this file, but now the
 # build steps are identical so we just have one script.
-# Was previously combined with osx-setup-opam-for-release.sh
+
+# Note that this script runs from a self-hosted CI runner which
```
Comment (aryx) [inline scripts/osx-setup-for-release.sh]: I'm not sure we still run self-hosted runners actually ...

File: src/engine/Xpattern_match_regexp.ml [hunk 10]
```diff
@@ -19,8 +19,48 @@ module MV = Metavariable
...
+  exec_all_noerr : 'regex -> string -> 'substring array;
+  get_substring : 'substring -> int -> string;
+  get_substring_ofs : 'substring -> int -> int * int;
+  names : 'libregex -> string array;
+  num_of_subs : 'substring -> int;
+  get_named_substring_ofs : 'libregex -> string -> 'substring -> int * int;
+  get_named_substring : 'libregex -> string -> 'substring -> string;
+}
+
+let pcre2_regex_functions =
```
Comment (aryx) [inline src/engine/Xpattern_match_regexp.ml:37]: can you add a comment why we need both pcre and pcre2 here? I thought the whole engine was switching to
pcre2 except spacegrep/aliengrep, so why we still need pcre here?

**Review by aryx — APPROVED** [review]
Supportive of this change if semgrep on macos is indeed working when dynamically
linked with libpcre2-8.

Please add a comment in osx-setup-...sh saying "Note that we are dynamically linking lippcre2 because it's complicated to use the remove trick here because blabla opam/git and because anyway it's available by default in macos in /usr/lib/ because blabla.
Note that we should not use this remove .dylib trick but statically linking on macos is hard, see andre-recent-link here."

**Comment by aryx** [comment]
Nice work. I know it was a huge pain. Thx.

File: Makefile (hunk 1 above)
Comment (kopecs) [reply Makefile]: Ah great; think I had just erred on the side of caution with a merge conflict.

File: changelog.d/scrt-467.changed (hunk 3 above)
Comment (kopecs) [reply changelog.d/scrt-467.changed:4]: > what [\w-.] meant? any word charater or '-' or '.'?

Previously, yes. But PCRE2 forbids these instances which look like ranges (e.g. `[a-z]`) but aren't valid as ranges.

> and what [\w_.] mean? any word character or '_' or '.'?

Good catch; the `[\w_.]` is a typo. It should be `[\w.-]` (moving the `-` to the end, so it can't be misunderstood as a range).

File: libs/aliengrep/Match.ml (hunk 6 above)
Comment (kopecs) [reply libs/aliengrep/Match.ml:8]: Per @mjambon's review on https://github.com/semgrep/semgrep/pull/9919

File: libs/spacegrep/src/lib/Comment.ml (hunk 8 above)
Comment (kopecs) [reply libs/spacegrep/src/lib/Comment.ml:8]: Ok.

File: scripts/osx-setup-for-release.sh (hunk 9 above)
Comment (kopecs) [reply scripts/osx-setup-for-release.sh]: This is just a revert of a prior commit. If you have up to date info to update it with just let me know.

File: src/engine/Xpattern_match_regexp.ml (hunk 10 above)
Comment (kopecs) [reply src/engine/Xpattern_match_regexp.ml:37]: Yes, it is because spacegrep/aliengrep use this. I can add a note.

File: libs/commons/Pcre2_.ml (hunk 7 above)
Comment (kopecs) [reply libs/commons/Pcre2_.ml:1]: Sure.

**Comment by kopecs** [comment]
Okay, based on some issues trying again with static linking but with dune flags rather than our hacky approach. Once I get something working I will split into another PR so this can be a single-issue pcre->pcre2 upgrade.

File: scripts/osx-setup-for-release.sh (hunk 9 above)
Comment (kopecs) [reply scripts/osx-setup-for-release.sh]: In any event, I am moving that revert commit + the other build changes to a separate PR here: https://github.com/semgrep/semgrep/pull/10122, so if you have a change just let me know in a review there. I am going to drop this and the other build debugging commits from this PR.

File: src/engine/Xpattern_match_regexp.ml (hunk 10 above)
Comment (mjambon) [reply src/engine/Xpattern_match_regexp.ml:37]: only aliengrep but yes.

File: dune-project (hunk 4 above)
Comment (kopecs) [reply dune-project]: Re-tagged and bumped to 7.5.3.

File: changelog.d/scrt-467.changed (hunk 3 above)
Comment (kopecs) [reply changelog.d/scrt-467.changed:4]: Fixed in rebase.
